### PR TITLE
perp positions changes

### PR DESCRIPTION
**Pull request by tinystarinagalaxy**

### CODE REVIEW

**Review by DimaCantemir — APPROVED** [review]

File: proto/api.proto [hunk 1]
```diff
@@ -367,6 +367,40 @@ service Api {
...
+  }
+
+  rpc GetPerpPositions(GetPerpPositionsRequest)  returns (GetPerpPositionsResponse) {
+    option (google.api.http) = {
+      get: "/api/v1/trade/perp-positions"
+    };
+    option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_operation) = {
+      description: "Returns the list current perp positions";
+      summary: "PERP positions";
+      tags: ["Trade", "Universal"];
```
Comment (aspin) [inline proto/api.proto]: Think this should be Orderbook, not Universal? Also, probably worth adding a "Perp" tag too

File: proto/api.proto [hunk 2]
```diff
@@ -367,6 +367,40 @@ service Api {
...
+  }
+
+  rpc PostClosePerpPositions(PostClosePerpPositionsRequest)  returns (PostClosePerpPositionsResponse) {
+    option (google.api.http) = {
+      get: "/api/v1/trade/close-perp-positions"
+    };
+    option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_operation) = {
+      description: "closes current perp positions";
+      summary: "PERP positions";
+      tags: ["Trade", "Universal"];
```
Comment (aspin) [inline proto/api.proto]: same comment ehre

File: proto/api.proto [hunk 3]
```diff
@@ -633,8 +667,6 @@ enum Side {
 enum OrderType {
   OT_MARKET = 0;
   OT_LIMIT = 1; // MARKET and LIMIT are mutually exclusive
-  OT_IOC = 2; // immediate or cancel
```
Comment (aspin) [inline proto/api.proto:636]: don't think we want to delete these –– they're valid for Serum/Openbook I think

File: proto/api.proto [hunk 4]
```diff
@@ -1109,4 +1142,64 @@ message GetPricesStreamResponse {
   TokenPrice price = 2;
 }
 
+message GetPerpPositionsRequest {
```
Comment (aspin) [inline proto/api.proto:1140]: minor: can you have request + response right next to each other in the file? Bit of work to scroll down to find `GetPerpPositionsResponses`

File: proto/api.proto [hunk 5]
```diff
@@ -1109,4 +1142,64 @@ message GetPricesStreamResponse {
...
+  Project project = 1;
+  string ownerAddress = 2;
+  repeated common.PerpContract contracts = 3;
+}
+
+message PostClosePerpPositionsResponse {
+  repeated string transactions = 1;
+}
+
+message GetCurrentPerpPosition {
```
Comment (aspin) [inline proto/api.proto]: naming: `PerpPosition`?

File: proto/api.proto [hunk 6]
```diff
@@ -1109,4 +1142,64 @@ message GetPricesStreamResponse {
...
+  common.PerpContract contract = 1;
+
+  double contractVolume = 2;
+  double volumeAvailable = 3;
+  double volumeInOrder = 4;
+
+  double positionMargin = 5;
+  string positionSide = 6; // long, short
```
Comment (aspin) [inline proto/api.proto]: probably should define an enum here

File: proto/api.proto (hunk 3 above)
Comment (aspin) [reply proto/api.proto:636]: actually, whats with the duplication with common.proto?

**Review by aspin — CHANGES_REQUESTED** [review]

File: proto/api.proto (hunk 2 above)
Comment (tinystarinagalaxy) [reply proto/api.proto]: added "Perp", "Orderbook" tags

File: proto/api.proto (hunk 1 above)
Comment (tinystarinagalaxy) [reply proto/api.proto]: added "Perp", "Orderbook" tags

File: proto/api.proto (hunk 3 above)
Comment (tinystarinagalaxy) [reply proto/api.proto:636]: yep, this is a leftover now. deleted this. moved OrderType to common.proto  and added new orderTypes.

File: proto/api.proto (hunk 5 above)
Comment (tinystarinagalaxy) [reply proto/api.proto]: renamed

File: proto/api.proto (hunk 4 above)
Comment (tinystarinagalaxy) [reply proto/api.proto:1140]: moved

File: proto/api.proto (hunk 6 above)
Comment (tinystarinagalaxy) [reply proto/api.proto]: already had common.PerpPositionSide, replaced string with common.PerpPositionSide

File: proto/api.proto [hunk 7]
```diff
@@ -367,6 +367,40 @@ service Api {
...
 
+  // perp endpoints
+  rpc PostPerpOrder(PostPerpOrderRequest)  returns (PostPerpOrderResponse) {
+    option (google.api.http) = {
+      post: "/api/v1/trade/perp/order"
+    };
+    option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_operation) = {
+      description: "post a perp order";
+      summary: "perp order";
+      tags: ["Trade", "Orderbook"];
```
Comment (aspin) [inline proto/api.proto]: Perp tag here too

File: proto/api.proto [hunk 8]
```diff
@@ -367,6 +367,40 @@ service Api {
...
+    };
+  }
+
+  rpc GetPerpPositions(GetPerpPositionsRequest)  returns (GetPerpPositionsResponse) {
+    option (google.api.http) = {
+      get: "/api/v1/trade/perp-positions"
```
Comment (aspin) [inline proto/api.proto]: `trade/perp/positions` I think

File: proto/api.proto [hunk 9]
```diff
@@ -367,6 +367,40 @@ service Api {
...
+    option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_operation) = {
+      description: "Returns the list current perp positions";
+      summary: "PERP positions";
+      tags: ["Trade", "Perp", "Orderbook"];
+    };
+  }
+
+  rpc PostClosePerpPositions(PostClosePerpPositionsRequest)  returns (PostClosePerpPositionsResponse) {
+    option (google.api.http) = {
+      get: "/api/v1/trade/close-perp-positions"
```
Comment (aspin) [inline proto/api.proto]: `/trade/perp/close-positions` I think

File: proto/api.proto (hunk 5 above)
Comment (aspin) [reply proto/api.proto]: I think also get rid of the `Get` in front of the name. This doesn't have to just be part of a `Get` requests and is a more general datatype

**Review by aspin — COMMENTED** [review]
couple more suggestions. good when these are done though

File: proto/api.proto [hunk 10]
```diff
@@ -1137,14 +1137,14 @@ message GetPricesStreamResponse {
...
-message GetPerpPositionsRequest {
+message PerpPositionsRequest {
   Project project = 1;
   string ownerAddress = 2;
   string accountAddress = 3;
   repeated common.PerpContract contracts = 4;
 }
 
-message GetPerpPositionsResponse {
+message PerpPositionsResponse {
```
Comment (aspin) [inline proto/api.proto]: think you misunderstood my comment here. These messages should have the `Get` in front of it. `GetPerpPosition` (field #3 here) is the one that shouldn't

Comment (aspin) [reply proto/api.proto]: the messages that match the RPC method (e.g. `GetPerpPositions(PerpPositionsRequest)  returns (PerpPositionsResponse)`) should always have the `Get` prefix. It's jsut the data types within the `Request`/`Response` objects that should not

**Review by aspin — APPROVED** [review]
take a look at the last comment then LGTM

**Review by aspin — APPROVED** [review]

File: proto/api.proto (hunk 10 above)
Comment (tinystarinagalaxy) [reply proto/api.proto]: fixed this

**Pull request merged**